Decouple CodeClimate EsLint and normal EsLint

FSVetaz opened the pull request:
Decouple codeclimate's eslint with our own eslint. CodeClimate uses it's own dependencies. Those dependencies were making us behind on our deps and were causing dependency conflicts (you have to do an npm i --force on master for install to work due to conflicts).

All packages are up-to-date now except for eslint-plugin-no-autofix. Not sure what we want to do with that because it was commented out for v10.

Updated CC to not lint in jest files:
eslint-8 channel has a dep on eslint-plugin-jest@26 which uses jest 27. eslint-plugin-jest@27 uses jest 29. Seems like it won't lint properly if it is depending on a wrong jest version. Also, not sure who codeclimate-lints their test files.
The local eslint will lint jest files just fine though.

I looked in fs-webdev and saw that no one is using the jsdoc config. Not sure if we want to keep that file or not. Looks like Clif wanted that so we will keep for now.

Typescript rules are added by default and it is only for the typescript files. So consumers don't need to worry about adding @fs/eslint-config-frontier-react/typescript to their extends.

We support both eslint@7 and eslint@8 now.

I didn't change anything for codeclimateEslintRules 4-10 except just pretterize the files. So, no need to look at those files.

Updated to use @babel/eslint-parser because babel-eslint is deprecated.
added es2021 env. es6 is still added in final config.
prettier has a default of traillingComma as es5, so it was clutter to have it deffine that.

Because all the react stuff is out of the es6 file, it is now possible for non-react repos to use this config by doing this:
"extends": [
    "@fs/eslint-config-frontier-react/es6",
    "@fs/eslint-config-frontier-react/typescript",
    "@fs/eslint-config-frontier-react/jest",
    "@fs/eslint-config-frontier-react/prettierSetup"
  ],

combined test-helpers with jest

_____ _____ _____ _____

UPDATED RULES:

`no-warning-comments` -- looks for todo, fixme, or hack anywhere in the comment now rather than todo, fixme, xxx at the start of a comment.

`import/no-extraneous-dependencies` -- setupTests file is now considered a devdep file.

`import/extensions` -- changed [ "error", "ignorePackages", { "js": "never", "mjs": "never", "jsx": "never" } ] -> ['error', 'ignorePackages', { js: 'never', jsx: 'never', ts: 'never', tsx: 'never' }] so that typescript files don't need .ts or .tsx extensions on them when importing; the import/extensions has been turned back on for ts files.

`import/prefer-default-export` -- turn off for mock JS/TS files

`react/jsx-no-bind` -- now a warning rather than an error. https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/jsx-no-bind.md#react-hooks is the way to fix it in code.

`react/no-unknown-property` -- turned off so that developers can use custom properties for css rules and props that parents use (like growX, columns). Note typescript, outside of eslint, checks to make sure props are added on correctly and if it is a custom prop that will be used for css rules, it requires a hyphen. A globals.d.ts file can be used to allow growX and such to be used on children. See the globals-react.d.ts file in zion.


`react/function-component-definition` --  turned off for now for fawg to determine what we want to do with it. It is a new rule that got added by eslint-plugin-react (https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/function-component-definition.md)

`react/jsx-props-no-spreading` -- new rule from eslint-plugin-react that I turned off because we spread overlay props all the time (https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/jsx-props-no-spreading.md)

`react/jsx-no-useless-fragment` -- new rule from eslint-plugin-react where allowExpressions is true because otherwise it will conflict with TS  (see https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/jsx-no-useless-fragment.md#allowexpressions)

`@typescript-eslint/no-use-before-define` -- from off to error but does not check functions because functions are hoisted so it is safe (https://eslint.org/docs/latest/rules/no-use-before-define#options)

`@typescript-eslint/no-empty-function` -- as error except for arrow functions to allow for empty functions when mocking functions or making them the default for a provider.

`@typescript-eslint/consistent-type-imports` -- error instead of warn; this separates type and non-type imports and it is automatically fixable. (For example makes this change automatically:
import React, { useState, useCallback, useRef, ReactNode } from 'react'

To 

import type { ReactNode } from 'react'
import React, { useState, useCallback, useRef } from 'react'
)

`@typescript-eslint/no-non-null-assertion` allow null asserting in test files because we know the real mock types here.

joeycozza commented on index.js:
```diff
@@ -1,12 +1,11 @@
-    // '@fs/eslint-config-frontier-react/noFixRules',
-    '@fs/eslint-config-frontier-react/dont-need-lodash',
-    '@fs/eslint-config-frontier-react/prettierSetup',
+    './react',
```
I have a faint memory that I used to do it this way, but that it messes things up if I tried to do npm linking (for testing purposes) in zion/react apps. 

I could be wrong though.

FSVetaz replied:
maybe. linking never seems to work right with me. I tested by installing from the github branch (npm i --sav-dev github:fs-webdev/eslint-config-frontier-react#no-codeclimate-dep) than from artifactory

joeycozza commented on index.js:
```diff
@@ -1,12 +1,11 @@
-    '@fs/eslint-config-frontier-react/prettierSetup',
+    './react',
+    './es6',
+    './typescript',
```
Again, I can't pin point the exact issue, but I'm pretty sure that when we defaulted to using typescript all the time, that some annoying/unwanted behavior was occurring. It is possible that we didn't have things setup for typescript to only run .ts/.tsx files, but I'd want to make sure this was testing/verified in a non typescript react app.

FSVetaz replied:
yep I tested in all  tw-blue apps which includes 2 typescript apps and 3 js apps and had no ts specific issues for the js apps.

joeycozza commented:
So a couple of points.

1. Typically rule changes go through a quick discussion/approval in FAWG since essentially all frontend teams use these rules.
2. I don't think that we can really decouple local eslint and CC eslint. Which lint rules take priority? We don't want to be in a position where the code is lint free locally, but CC reports that lint exists, or vice versa. CC is important since it is used in the quality metrics 
3. I'm fine with removing jsdoc, I was asked to add it by Clif, and since it doesn't really hurt anything, I added it as optional.

FSVetaz commented:
> So a couple of points.
> 
> 1. Typically rule changes go through a quick discussion/approval in FAWG since essentially all frontend teams use these rules.
> 2. I don't think that we can really decouple local eslint and CC eslint. Which lint rules take priority? We don't want to be in a position where the code is lint free locally, but CC reports that lint exists, or vice versa. CC is important since it is used in the org's quality metrics and some people's ATI's etc. (I believe @marybeshaw mentioned that to me once)
> 3. I'm fine with removing jsdoc, I was asked to add it by Clif, and since it doesn't really hurt anything, I added it as optional.

I'll come up with a list that has the major rule changes that are introduced either because of an update of a dep or because of a rule that I updated (like no-warning-comments).
local eslint should take priority. CC eslint rules can be invalidated when an admin of a repo deems it is invalid. It won't be exactly the same as local eslint. We do want them to be close as possible though. 
Ah ok, we can keep the jsdoc in there for now.

FSVetaz commented:
Looks like this repo is missing some password? 
![image](https://user-images.githubusercontent.com/77301861/210011858-f43ddb43-76d0-4b91-ac7b-a8886623ab0f.png)

joeycozza commented:
> Looks like this repo is missing some password? ![image](https://user-images.githubusercontent.com/77301861/210011858-f43ddb43-76d0-4b91-ac7b-a8886623ab0f.png)

the .npmrc that you added was messing things up for some reason that I don't understand. I deleted it and travis was happy again.

joeycozza reviewed (changes requested):
I meant to just comment, not approve. So I am "requesting changes". The "changes" is that this gets discussed in our next FAWG meeting. I'll send you an invite to that.

FSVetaz commented:
example of invalidating a rule because it is invalid
![image](https://user-images.githubusercontent.com/77301861/210094032-62a270b9-61fd-4dfe-8eae-74b2d97dc16d.png)

eslint-plugin-react-hooks v1 doesn't support optional chaining but eslint-plugin-react-hooks v4 does (which is the version we use locally). The autofix for v1 actually autofixes to focusPerson.id instead of focusPerson?.id, so that is possibly bad. So it is good that our localeslint has v4, and we just have to put up with invalidating that on CC, but we need to keep the check on for reporting when there is a real issue.

joeycozza commented:
Going to close this PR since we stopped supporting codeclimate in v11 from your more updated PR.

If there are additional rule changes you'd like to see, you can make a new PR and discuss it in FAWG with us.
